File: python/blur-optimized.py
import math
import random
import time

WIDTH = 500
HEIGHT = 500
SIZE = WIDTH * HEIGHT
RADIUS = 32

band = []

for i in range(0, SIZE):
    band.append(math.floor(random.randint(0, 255)))

print(band[0: 100], "...", SIZE - 100, "more items")

start = time.time()

firstNewBand = []

for h in range(0, HEIGHT):
    sumOfNumbers = 0
    cellCount = RADIUS + 1

    for j in range(0, RADIUS + 1):
        sumOfNumbers += band[WIDTH * h + j]

    firstNewBand.append(round(sumOfNumbers / cellCount))

    for w in range(0, WIDTH):
        if w > RADIUS:
            sumOfNumbers -= band[WIDTH * h + w - RADIUS - 1]
            cellCount -= 1
        if w < WIDTH - RADIUS:
            sumOfNumbers += band[WIDTH * h + w + RADIUS]
            cellCount += 1

        firstNewBand.append(round(sumOfNumbers / cellCount))


secondNewBand = list(range(0, SIZE))

for w in range(0, WIDTH):
    sumOfNumbers = 0
    cellCount = RADIUS + 1

    for j in range(0, RADIUS + 1):
        sumOfNumbers += firstNewBand[w + j * WIDTH]

    secondNewBand[w] = round(sumOfNumbers / cellCount)

    for h in range(0, HEIGHT):
        if h > RADIUS:
            sumOfNumbers -= firstNewBand[w + WIDTH * (h - RADIUS - 1)]
            cellCount -= 1

        if h < HEIGHT - RADIUS:
            sumOfNumbers += firstNewBand[w + WIDTH * (h + RADIUS)]
            cellCount += 1

        secondNewBand[w + WIDTH * h] = round(sumOfNumbers / cellCount)

elapsed = time.time() - start

print("Time taken:", elapsed, "seconds")
print(secondNewBand[0: 100], "...", SIZE - 100, "more items")

File: js/blur-optimized.js
const width = 500,
  height = 500,
  size = width * height,
  radius = 32;

let band = [];

for (let i = 0; i < size; i++) {
  band.push(Math.floor(Math.random() * 255));
}

console.log("Original:", band);

function main() {
  let start = Date.now();

  let firstNewBand = [];

  for (let h = 0; h < height; h++) {
    let sum = 0;
    let cellCount = radius + 1;

    for (let j = 0; j <= radius; j++) {
      sum += band[width * h + j];
    }

    firstNewBand.push(Math.round(sum / cellCount));

    for (let w = 1; w < width; w++) {
      if (w > radius) {
        sum -= band[width * h + w - radius - 1];
        cellCount--;
      }
      if (w < width - radius) {
        sum += band[width * h + w + radius];
        cellCount++;
      }
      firstNewBand.push(Math.round(sum / cellCount));
    }
  }

  let secondNewBand = [];

  for (let w = 0; w < width; w++) {
    let sum = 0;
    let cellCount = radius + 1;

    for (let j = 0; j <= radius; j++) {
      sum += firstNewBand[w + j * width];
    }

    secondNewBand[w] = Math.round(sum / cellCount);

    for (let h = 1; h < height; h++) {
      if (h > radius) {
        sum -= firstNewBand[w + width * (h - radius - 1)];
        cellCount--;
      }
      if (h < height - radius) {
        sum += firstNewBand[w + width * (h + radius)];
        cellCount++;
      }
      secondNewBand[w + width * h] = Math.round(sum / cellCount);
    }
  }

  let elapsed = (Date.now() - start) / 1000;

  console.log("Time taken:", elapsed, "seconds");
  console.log("Output:", secondNewBand);
}

for (let i = 0; i < 10; i++) {
  main();
}
